service/src/main/java/bio/terra/workspace/app/controller/FolderApiController.java

```diff
@@ -107,9 +107,9 @@ public ResponseEntity<ApiFolderList> listFolders(UUID workspaceId) {
   @Override
   public ResponseEntity<Void> deleteFolder(UUID workspaceId, UUID folderId) {
     AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
-    // Only owner should be able to perform the delete.
-    // TODO(PF-1965): Allow writer to delete too but have special handling for private resources.
-    workspaceService.validateWorkspaceAndAction(userRequest, workspaceId, SamWorkspaceAction.OWN);
+
+    workspaceService.validateWorkspaceAndAction(userRequest, workspaceId, SamWorkspaceAction.WRITE);
```

Comment:
What error does the user see?

I think it's worth it to provide a more helpful error message for this case. Otherwise, user will be confused. WDYT

- Here call FolderService.checkDeleteAction(folderId);
- FolderService.checkDeleteAction {
    if requester is owner, return
    // requester must be writer
    List<ControlledResource> controlledResourcesInFolder = ...
    if any controlled resource is private and requester isn't owner:
       throw new ForbiddenException("Unable to delete private resource %s in folder %s. Only workspace owners or resource owner can delete it.", resource.getName(), folder.getName());
 }

Also, I think it's nice to do access control in one place.

Reply:
This check is just to see if the user has WRITE action at all on the workspace. It is not saying anything about the folder. Maybe the comment applies later in the PR?

Reply:
i will wrap the private resource error with something more actionable.

But i think it makes sense to throw early here if the person is just a reader. And the error you don't have write permission on this workspace makes sense for a reader.

Do you want to fetch the resources and perform the check in the controller to have the all the auth in controller or is it OK to do the private resource check in the service? For the first case, the FolderService would have
`deleteFolder(workspaceId, folderId, controlledResources, referencedResources, userRequest)`

Reply:
Sorry, I meant to keep the current check and add a call to some method in folderService.

```
// Validates requester has at least writer role on workspace
workspaceService.validateWorkspaceAndAction(userRequest, workspaceId, SamWorkspaceAction.WRITE);

// Validates requester has permission to delete every resource in folder
folderService.validateDeleteAction();
```

(And the error message should print resource names, that's more helpful, and resource name is unique anyways.)

Reply:
folderservice.validateDeleteaction() would requires making multiple calls to folder and resource db to fetch all the resources. And folderService.deleteFolder would need to do that again, because i'm passing resources into the flight. so I don't want to do this action twice.

Reply:
Then just add a comment here. Normally access checking is done in the controller beginning, so it's nice to comment when that's not the case.

// If requester is writer and folder has private resources (not owned by requester), requester won't have permission to delete private resources. That access control check is done in folderService.XXX.